File: src/features/typing/data/go/find_median_from_data_stream.go
package main

import "sort"

type MedianFinder struct {
	sorted []int
}

func MedianFinderConstructor() MedianFinder {
	return MedianFinder{}
}

func (f *MedianFinder) AddNum(num int) {
	idx := sort.SearchInts(f.sorted, num)
	f.sorted = append(f.sorted, 0)
	copy(f.sorted[idx+1:], f.sorted[idx:])
	f.sorted[idx] = num
}

func (f *MedianFinder) FindMedian() float64 {
	n := len(f.sorted)
	if n%2 == 1 {
		return float64(f.sorted[n/2])
	}
	return float64(f.sorted[n/2-1]+f.sorted[n/2]) / 2
}

File: src/features/typing/data/java/find_median_from_data_stream.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class MedianFinder {
    private final List<Integer> sorted = new ArrayList<>();

    public MedianFinder() {}

    public void addNum(int num) {
        int idx = Collections.binarySearch(sorted, num);
        if (idx < 0) {
            idx = -(idx + 1);
        }
        sorted.add(idx, num);
    }

    public double findMedian() {
        int n = sorted.size();
        if (n % 2 == 1) {
            return sorted.get(n / 2);
        }
        return (sorted.get(n / 2 - 1) + sorted.get(n / 2)) / 2.0;
    }
}
